### PR TITLE
fix(hostnamegenerator): selectors validation and matching

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### Checklist prior to review

I noticed a bug when implementing hostnames for MeshMultiZoneService.
If we define this HostnameGenerator
```
spec:
  selectors:
    meshService: {}
```
It also selects MeshExternalServices.

It does not make sense to have one HostnameGenerator for MeshExternalService, MeshService, so we should require exact one field in Selectors.
Additionally, `MeshService` and `MeshExternalService` fields were not pointers so it was impossible to distinguish
```
spec:
  selectors: {}
```
from
```
spec:
  selectors:
    meshService: {}
```

I also added builders, but not samples because we are still unsure what good default values are.

- [X] [Link to relevant issue][1] as well as docs and UI issues --
- [X] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

File: pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/hostnamegenerator.go [hunk 1]
```diff
@@ -14,8 +14,8 @@ type NameLabelsSelector struct {
 }
 
 type Selector struct {
-	MeshService         LabelSelector      `json:"meshService,omitempty"`
-	MeshExternalService NameLabelsSelector `json:"meshExternalService,omitempty"`
+	MeshService         *LabelSelector      `json:"meshService,omitempty"`
+	MeshExternalService *NameLabelsSelector `json:"meshExternalService,omitempty"`
```
Comment (lukidzi) [inline pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/hostnamegenerator.go:18]: Should we change to `LabelSelector` ?

**Review by lukidzi — APPROVED** [review]
I assume we use `merge release to master` to sync the change?

File: pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate.go [hunk 2]
```diff
@@ -11,10 +11,26 @@ import (
...
+	var verr validators.ValidationError
+	selectorsDefined := 0
+	if selector.MeshService != nil {
+		selectorsDefined++
+	}
+	if selector.MeshExternalService != nil {
+		selectorsDefined++
+	}
+	if selectorsDefined != 1 {
+		verr.AddViolationAt(path, "exact one selector (meshService, meshExternalService) must be defined")
```
Comment (slonka) [inline pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate.go:29]: nitpick:

```suggestion
		verr.AddViolationAt(path, "exactly one selector (meshService, meshExternalService) must be defined")
```

File: pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate_test.go [hunk 3]
```diff
@@ -28,6 +30,29 @@ template: "{{ .Name[4 }}.mesh"
 type: HostnameGenerator
 name: route-1
 template: ""
+selector:
+  meshService: {}
+`),
+		ErrorCase("spec.selector empty",
+			validators.Violation{
+				Field:   `spec.selector`,
+				Message: `exact one selector (meshService, meshExternalService) must be defined`,
```
Comment (slonka) [inline pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate_test.go:39]:
```suggestion
				Message: `exactly one selector (meshService, meshExternalService) must be defined`,
```

File: pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate_test.go [hunk 4]
```diff
@@ -28,6 +30,29 @@ template: "{{ .Name[4 }}.mesh"
...
+			}, `
+type: HostnameGenerator
+name: route-1
+template: "{{ .Name }}.mesh"
+`),
+		ErrorCase("spec.selector has too many selectors",
+			validators.Violation{
+				Field:   `spec.selector`,
+				Message: `exact one selector (meshService, meshExternalService) must be defined`,
```
Comment (slonka) [inline pkg/core/resources/apis/hostnamegenerator/api/v1alpha1/validate_test.go:48]:
```suggestion
				Message: `exactly one selector (meshService, meshExternalService) must be defined`,
```

**Review by slonka — APPROVED** [review]
just nits

**Pull request merged**